src/status.rs

```diff
@@ -0,0 +1,126 @@
+use std::fmt;
+use std::process::ExitCode;
+
+use colored::Colorize;
+
+use crate::nixpkgs_problem::NixpkgsProblem;
+
+pub enum Status {
+    /// It's all green.
+    ValidatedSuccessfully,
+
+    /// The base branch is broken, but this PR fixes it. Nice job!
+    BranchHealed,
+
+    /// The base branch fails, the PR doesn't fix it, and the PR may also introduce additional
+    /// problems.
+    BranchStillBroken(Vec<NixpkgsProblem>),
+
+    /// This PR introduces the problems listed. Please fix them before merging, otherwise the base
+    /// branch would break.
+    ProblemsIntroduced(Vec<NixpkgsProblem>),
+
+    /// This PR introduces additional instances of discouraged patterns. Merging is discouraged but
+    /// would not break the base branch.
+    DiscouragedPatternedIntroduced(Vec<NixpkgsProblem>),
+
+    /// Some other error occurred.
+    Error(anyhow::Error),
+}
+
+impl Status {
+    fn fmt(&self, f: &mut fmt::Formatter, use_color: bool) -> fmt::Result {
+        let maybe_green = |s: &str| if use_color { s.green() } else { s.normal() };
+        let maybe_yellow = |s: &str| if use_color { s.yellow() } else { s.normal() };
+        let maybe_red = |s: &str| if use_color { s.red() } else { s.normal() };
```

Comment:
This will cause it to colorise the output even if it's e.g. just a pipe to a file. So I think I prefer the previous code using `NO_COLOR=1` to disable colors when testing, and https://github.com/NixOS/nixpkgs/blob/07c1a652be3a1dda9047f4d8545e17d12958a148/.github/workflows/check-by-name.yml#L113-L115 to force turn it on in CI.

Reply:
No, it's the same as before.

`nixpkgs-check-by-name` used to unilaterally colorize by calling the `.red`, `.yellow`, etc methods, and in the `colored` crate it would decide to _actually_ not color based on [this environment variable](https://github.com/colored-rs/colored/blob/master/src/control.rs#L98-L101).

Now, it'll only call those methods if `true` is passed to `use_color`. `ColoredStatus` does this -- see the end of the file. `main` uses `ColoredStatus`, and so it retains the original behavior. Tests, which use the `process` method, format a `Status`, which passes `false` to `use_color` in their display, bypassing colorization in this module itself.